src/sources/kubernetes.rs

```diff
@@ -0,0 +1,318 @@
+        .include
+        .push((LOG_DIRECTORY.to_owned() + r"*/*/*.log").into());
+
+    config.start_at_beginning = true;
+
+    // Filter out files that certainly don't have logs newer than now timestamp.
+    config.ignore_older = Some(10);
+
+    // oldest_first false, having all pods equaly serviced is of greater importance
+    //                     than having time order guarantee.
+
+    // CRI standard ensures unique naming.
+    config.fingerprinting = FingerprintingConfig::DevInode;
+
+    // Have a subdirectory for this source to avoid collision of naming its file source.
+    config.data_dir = Some(globals.resolve_and_make_data_subdir(None, kube_name)?);
+
+    let (file_send, file_recv) = mpsc::channel(1000);
+    let file_source = config
+        .build("file_source", globals, file_send)
+        .map_err(|e| format!("Failed in creating file source with error: {:?}", e))?;
+
+    Ok((file_recv, file_source))
+}
+
+fn parse_message() -> Result<impl FnMut(Event) -> Option<Event> + Send, String> {
+    // Transforms
+    let mut transform_json_message = transform_json_message();
+    let mut transform_cri_message = transform_cri_message()?;
+
+    // Kubernetes source
+    let atom_pod_uid = Atom::from("pod_uid");
+    let mut pod_data = HashMap::<Bytes, PodData>::new();
+    Ok(move |event: Event| {
+        let log = event.as_log();
+        if let Some(ValueKind::Bytes(pod_uid)) = log.get(&atom_pod_uid) {
+            // Fetch pod data
+            let data = pod_data.entry(pod_uid.clone()).or_default();
+
+            // Get/determine log format
+            let log_format = if let Some(log_format) = data.log_format {
+                log_format
+            } else {
+                // Detect log format
+                let log_format = if transform_json_message(event.clone()).is_some() {
+                    LogFormat::Json
+                } else if transform_cri_message.transform(event.clone()).is_some() {
+                    LogFormat::CRI
+                } else {
+                    error!(message = "Unknown message format");
+                    // Return untouched message so that user has some options
+                    // in how to deal with it.
+                    LogFormat::Unknown
+                };
+                // Save log format
+                data.log_format = Some(log_format);
+                log_format
+            };
+
+            // Parse message
+            match log_format {
+                LogFormat::Json => transform_json_message(event),
+                LogFormat::CRI => transform_cri_message.transform(event),
+                LogFormat::Unknown => Some(event),
+            }
+        } else {
+            None
+        }
+    })
+}
+
+fn transform_json_message() -> impl FnMut(Event) -> Option<Event> + Send {
+    let mut config = JsonParserConfig::default();
+
+    // Drop so that it's possible to detect if message is in json format
+    config.drop_invalid = true;
+
+    // Drop field
```

Comment:
do we need this comment?

Reply:
Yes, just so it's easier to deduce what fields are in the event at which point. But there is a better way to express it so I'll add that.